klippy/extras/bltouch.py

```diff
@@ -53,54 +59,96 @@ def __init__(self, config):
         self.gcode = self.printer.lookup_object('gcode')
         self.gcode.register_command("BLTOUCH_DEBUG", self.cmd_BLTOUCH_DEBUG,
                                     desc=self.cmd_BLTOUCH_DEBUG_help)
+        # Debugging code to report sensor pin state changes
+        ppins.reset_pin_sharing(pin_params)
+        buttons = self.printer.try_load_module(config, "buttons")
+        buttons.register_buttons([pin], self.sensor_debug_callback)
+    def sensor_debug_callback(self, eventtime, state):
+        logging.info("bltouch sensor @ %.6f is %s", eventtime, state)
     def _build_config(self):
         kin = self.printer.lookup_object('toolhead').get_kinematics()
         for stepper in kin.get_steppers('Z'):
             stepper.add_to_endstop(self)
-    def send_cmd(self, print_time, cmd):
-        self.mcu_pwm.set_pwm(print_time, Commands[cmd] / SIGNAL_PERIOD)
-    def test_sensor(self):
-        if not self.test_sensor_pin:
-            return
+    def handle_connect(self):
+        self.sync_mcu_print_time()
+        try:
+            self.raise_probe()
+        except homing.EndstopError as e:
+            raise self.printer.config_error(str(e))
+    def sync_mcu_print_time(self):
+        curtime = self.printer.get_reactor().monotonic()
+        est_time = self.mcu_pwm.get_mcu().estimated_print_time(curtime)
+        self.next_cmd_time = max(self.next_cmd_time, est_time + MIN_CMD_TIME)
+    def sync_print_time(self):
         toolhead = self.printer.lookup_object('toolhead')
         print_time = toolhead.get_last_move_time()
-        if print_time < self.next_test_time:
-            self.next_test_time = print_time + TEST_TIME
-            return
-        # Raise the bltouch probe and test if probe is raised
-        self.send_cmd(print_time, 'reset')
-        home_time = print_time + self.pin_move_time
-        self.send_cmd(home_time, 'touch_mode')
-        self.send_cmd(home_time + MIN_CMD_TIME, None)
-        # Perform endstop check to verify bltouch reports probe raised
+        if self.next_cmd_time > print_time:
+            toolhead.dwell(self.next_cmd_time - print_time)
+        else:
+            self.next_cmd_time = print_time
+    def send_cmd(self, cmd, duration=MIN_CMD_TIME):
+        self.mcu_pwm.set_pwm(self.next_cmd_time, Commands[cmd] / SIGNAL_PERIOD)
+        # Translate duration to ticks to avoid any secondary mcu clock skew
+        mcu = self.mcu_pwm.get_mcu()
+        cmd_clock = mcu.print_time_to_clock(self.next_cmd_time)
+        cmd_clock += mcu.seconds_to_clock(max(duration, MIN_CMD_TIME))
+        self.next_cmd_time = mcu.clock_to_print_time(cmd_clock)
+        return self.next_cmd_time
```

Comment:
It would only apply to those running a bltouch on a multi-mcu setup and that have the bltouch on a secondary mcu.  It would be rare for it to cause a problem even in that case, but I felt it was worthwhile to get the timing correct.